### DISCUSSION TITLE
Change classic mod multiplier to 0.98x (for osu!std)

### DISCUSSION THREAD

**Discussion by sz-da** [ppy/osu#30825] [Ideas]
Currently, the 0.96x score multiplier does not accurately represent the majority of beatmaps when it comes to slider acc.

Using a random sample of 700 beatmaps and the top 10 (stable leaderboard) scores on them, I calculated the scores with and without slider acc, and the ratio between the two.
[spreadsheet](https://docs.google.com/spreadsheets/d/1QSLpsTq-_Z57cNjw2Vy6hb1JAhxXHPdTOVAPLmYKBS4/edit?usp=sharing)

Median and average of the ratios:
Median: `0.982`
Average: `0.971`

Graphing the ratios in a histogram shows over 70% of scores have a ratio of > 0.96

<img src="https://github.com/user-attachments/assets/93b433de-aedd-4926-9829-c5e83539152e" alt="score comparison" width="80%">

\
Graphing the median ratio per beatmap also gives a similar result, with the majority of beatmaps having a ratio between 0.97 and 1

<img src="https://github.com/user-attachments/assets/a2adc10e-bfea-4d6e-8eed-b3e9e8e07f29" alt="score comparison (grouped by beatmap)" width="80%">

\
From this data, I suggest changing the classic mod score multiplier to 0.98x to better balance CL scores with non-CL scores.

<br/>

### How the scores were calculated
Circleguard was used to calculate the hitcounts with and without slider acc. The hitcounts were then plugged into the total score equation to get the score.

Notes:
- All sliderticks were assumed to be hit
- ~~All scores were assumed to be FCs~~
- Circleguard isn't 100% accurate, but it's accurate enough where any errors will have very little effect on the results presented here.

**Some extra stats:**
Lowest ratio: `0.520` [score link](https://osu.ppy.sh/scores/264217786)
Highest ratio: `1.020` [score link](https://osu.ppy.sh/scores/1693197225)
Lowest median ratio: `0.859` [beatmap link](https://osu.ppy.sh/beatmapsets/1496646#osu/3067984)
Highest median ratio: `1.007` [beatmap link](https://osu.ppy.sh/beatmapsets/673878#osu/1430822)

Edit:
- fixed values from using incorrect base score for classic mod sliderends (150 -> 10)
- recalculated values with correct combo progress

[Link](https://github.com/ssz7-ch2/classic-mod-multiplier) to script used to calculate the score ratios

**Comment by Zyfarok** [comment]
Thank you for those graphs !

Maybe 0.98 works better for "the majority of scores" (Edit: actually I didn't realise this was only based on top 10, this is a problem...) but:
- Any multiplier that is not x1.00 will always be overly punitive for SS scores (and near SS scores)
- Any multiplier that is not <0.9 will always be overly generous to low acc scores on most maps

And more importantly, if the multiplier is not <0.97, people will have too much incentive to enable classic mod (especially on slider-heavy maps) since they will usually get better scores with it, but if the multiplier is not >0.99, the top of the leaderboard won't make sense, and thus there's no compromise that truly works.

This is why the solution of changing the score conversion to include an estimation of the slider-accuracy was proposed, so that stable scores and lazer scores can be fairly compared, and I understand that having a simple multiplier for mods is desirable because it's easier to understand, but here we're dealing with a situation that requires careful handling since people care about these historical scores, and this complexity is the only way to provide the fairness that players want here. So, I hope the devs change their mind on that topic.
(Note: I don't care how lazer's CL scores are treated and whether they are treated like classic scores, I'm only concerned about stable score here)

**Reply by Zyfarok** [reply]
Actually, from the 1st graph you can see that:
- for more than 25% of scores, the ideal multiplier should be <0.96 (1st quartile < 0.96)
- for more than 25% of scores, the multiplier should be >0.99 (3rd quartile > 0.99)

That's a lot of under- or over-evaluated scores, no matter which multiplier you pick: >25% will be off by 1.5% or more compared to a fair conversion. (Edit: And this data is only for top 10 scores!)

**Comment by Zyfarok** [comment]
On another note, in case the suboptimal solution of using a mod multiplier to compare classic/lazer is forced to remain, I believe the median is not a good metric to pick the multiplier:
The average (~0.97 here) minimizes the squared-error and seems a better choice for that reason IMO. Also, as I stated above you don't want to incentivize picking the CL mod to get more score easily, so you might want to go lower than the average (which is why the current 0.96 is not that bad).
The only argument for a multiplier above 0.97 is that high-acc/SS scores are more visible in the leaderboard thus they might have more weight in the decision (people care more about them ?), but I'm not sure if it is a good argument...

Edit: also, why only use top 10 scores in leaderboards ? What happens to lower acc scores ?

**Reply by WitherFlower** [reply]
> The only argument for a multiplier above 0.97 is that high-acc/SS scores are more visible in the leaderboard thus they might have more weight in the decision

This is such an understatement. 99% of the leaderboards in osu are **absolutely dominated** by high-accuracy/SS scores. I'd rather have a high multiplier that accommodates for this 99% instead of having a middle-ground that is a-little-bit-bad-in-all-cases.

**Reply by Zyfarok** [reply]
Not on maps that are actually hard (with or without mods), which in my opinion is the only place where the top of the leaderboard actually matters and is interesting to look at, and also not in more restricted leaderboards like friends leaderboard. (unless you're a score farmer with score farming friends and only play 3* maps...)

**Reply by sz-da** [reply]
> Not on maps that are actually hard (with or without mods), which in my opinion is the only place where the top of the leaderboard actually matters and is interesting to look at, and also not in more restricted leaderboards like friends leaderboard. (unless you're a score farmer with score farming friends and only play 3* maps...)

Was interested in how the score ratio related to SR, so I made a graph for it.
The star ratings are rounded down, so 3.0 means `3.0 <= sr < 3.5`

<img src="https://github.com/user-attachments/assets/a76bcb65-186a-44d0-a7d3-3f3c31ad92c6" alt="Grouped by Star Rating (NM)" width="80%">

\
Seems like 2.5 - 3.5* maps have the lowest ratio, though it does make sense if you think about it.
Leaderboards for these maps often have HDHRDT AR10.5+ scores in the top 10. And the players playing these maps usually can't sightread AR10.5+, meaning they have to memorize parts of the maps. And since stable doesn't have slider acc, no one bothers to memorize the sliders, leading to a lower score ratio. At least that's my hypothesis on why it happens.

And here's a graph comparing score ratio with AR.

<img src="https://github.com/user-attachments/assets/fab55a7c-58d8-4e1b-96f8-9330a3366e02" alt="Grouped by AR" width="80%">

\
If we filter for scores with `ar >= 10.5`, the avg ratio drops to `0.925` and median ratio drops to `0.943`


Another thing to consider is that FL scores on average have a lower score ratio. Probably due to similar reasons as high AR scores, where players don't bother to memorize the sliders, leading to lower score ratios.
Though for most of these FL scores, I don't think it's that much harder to accuracy with slider acc, since the map is like 2*

If we filter for scores with `fl >= 0.01`, the avg ratio drops to `0.964` and median ratio drops to `0.977`

On the other hand, if we filter for scores with `fl == 0` and `ar < 10.5`, the avg ratio increases to `0.978` and median ratio increase to `0.985`

So 0.98x is fairly balanced for scores that aren't high AR HDHRDT scores or FL scores.

**Reply by Zyfarok** [reply]
Alright, my guess might have been wrong on that point. Thank you for the per-sr data. 

Someone I discussed with made me realize that other aspects of lazer vs classic partially balance it out: in particular, slider-ends giving much more additional base-score in lazer means the maximum total base-score (used to compute accuracy) is higher, thus every 100 will cost less accuracy in lazer than classic/stable. This probably partially explains why higher SR does not imply lower "ideal" multiplier.

However, I have big concerns about the fact that you base your multiplier solely on the top 10 on those beatmaps, instead of all scores in the leaderboard. What does it look like for lower scores ? Based on the point above, it should be fine, but I would want to be 100% sure that classic doesn't end-up being OP on low acc scores or something. The people trying to SS maps with HDHRDT(FL) are a tiny minority, I don't think CL should be solely tuned for them.

**Reply by sz-da** [reply]
The top 10 was mainly used since getting the replays does take some time. So I limited it to the top 10 since people care more about them.

Here's the stats for a random sample of 360 beatmaps and 10 random scores from the top 11-50 for each beatmap:
Median: `0.983`
Average: `0.973`

And with FL scores and high AR scores filtered out:
Median: `0.984`
Average: `0.975`

Slightly lower average compared to the top 10, though the median is basically the same.

The star rating graph also looks pretty similar.

<img src="https://github.com/user-attachments/assets/1eed21f0-ac63-41d7-a46c-361f0cc35c41" alt="Grouped by Star Rating (NM)" width="80%">

**Reply by Zyfarok** [reply]
Sorry I should have clarified, but if those maps have hundreds of plays, there's still a huge bias when selecting the top 50.
What I would want to see is the graph for some plays selected truly randomly without limiting to the top 50, OR otherwise on maps with less than 50 plays (some qualified maps?)

**Reply by sz-da** [reply]
The api only shows up to the top 100, so don't think that would change much for ranked maps.

Here's the stats for 100 qualified beatmaps and 10 random scores from the top 50:
Median: `0.930`
Average: `0.901`

Graph for the score ratios

<img src="https://github.com/user-attachments/assets/3638cb11-611e-497c-8922-fe50bce6800b" alt="score comparison" width="80%">

\
And graph for the star ratings

<img src="https://github.com/user-attachments/assets/732659d4-11ff-410f-84bd-71875daaa884" alt="Grouped by Star Rating (NM)" width="80%">

\
For beginner players playing maps well above their comfortable zone, CL does give a huge score advantage. Though for these scores, changing the multiplier from 0.96 to 0.98 won't make much of a difference.

**Reply by Rekunan** [reply]
If we want low acc scores, how about using leaderboards of way higher SR maps? Like 10*+.

**Reply by Zyfarok** [reply]
> The api only shows up to the top 100, so don't think that would change much for ranked maps.

The reason I'm asking for that is that the order also matters for local, country and friends leaderboards and there you definitely expect lower accuracy scores.

CL being advantageous to beginners is annoying but it's not a big issue IMO. What I'm more concerned about is CL being advantageous on 5*+ map for most players.

**Reply by sz-da** [reply]
> What I'm more concerned about is CL being advantageous on 5*+ map for most players.

Looking through some of the scores on the qualified 5*+ maps, the ratio is closer to 0.9 for scores that are a B or C rank, and closer to 0.95-1 if the acc is 95%+. So CL will be advantageous if the player can't play the map well.
Not sure how much of a concern this is, since this is also a problem for the current 0.96x multiplier, just to a lesser extent.

**Reply by Zyfarok** [reply]
btw @ssz7-ch2, first thx again for the insightful data.

Just to be sure:
You say you took into account that slider-ends base-value is now 150 instead of 10 in stable.
But do you also take into account:
- that (AFAIK) stable computes acc differently (using count of 300/100/50/miss only, ignoring ticks/ends base-values) than lazer (which uses base-score, and thus accounts for slider-ticks/ends base-values) ? 
- that in stable, judgements on the slider head is impacted by missed slider-ends/ticks ?
- that in lazer scoring, the ComboScore portion of the score is also impacted by the acc ?

Overall, it would be nice if you could provide us your code to allow us to double-check how you compute things. (be careful to NOT share your API key)

**Reply by Zyfarok** [reply]
Also, the above makes me actually question how lazer's classic scores and stable scores can actually be represented with absolutely no distinction and the same mod multiplier. Stable scores should probably be treated with a different multiplier than classic because of that.

**Reply by sz-da** [reply]
> * that (AFAIK) stable computes acc differently (using count of 300/100/50/miss only, ignoring ticks/ends base-values) than lazer (which uses base-score, and thus accounts for slider-ticks/ends base-values) ?

Stable acc wasn't used for the calculations. I should've clarified that I am recalculating the stable scores as if they were set on lazer for both with and without slider acc, not the stable -> lazer converted score. The converted score can be a bit off sometimes, so probably not a good idea to use it in the comparisons.

> * that in stable, judgements on the slider head is impacted by missed slider-ends/ticks ?

Missed sliderends are accounted for, but all sliderticks are assumed to be hit (not really an issue except for really crazy slider maps)

> * that in lazer scoring, the ComboScore portion of the score is also impacted by the acc ?

Yes. This part of the total score calculation was copied 1 to 1 from lazer.


I will look into uploading the code in a bit. Need to clean it up a bit and add some comments to explain the code.

**Reply by sz-da** [reply]
> Overall, it would be nice if you could provide us your code to allow us to double-check how you compute things. (be careful to NOT share your API key)

[Link to code](https://github.com/ssz7-ch2/classic-mod-multiplier)
And also the code for calculating combo portion [here](https://github.com/ssz7-ch2/circlecore/blob/15e880c90d472ea3c3afb8e916ec598d434bed1a/circleguard/investigations.py#L334-L413)

**Reply by sz-da** [reply]
@Rekunan 
> If we want low acc scores, how about using leaderboards of way higher SR maps? Like 10*+.

Grabbed like 300 scores from 10*+ maps and the average was `0.975`. A lot higher than what you would expect for low acc scores, but it kinda makes sense. Most 10*+ maps have a lot of circles and often have hard to hit sliderends, which reduces the impact that slider 100s have.

**Reply by Zyfarok** [reply]
@ssz7-ch2 
> Stable acc wasn't used for the calculations

That's an issue. Nearly nobody cares about the fairness of lazer classic vs lazer non-classic, as long as lazer-classic is not broken.
What most people care about is (historical) stable score vs lazer scores (without classic). I know there's a conversion involved on the ComboScore portion, but it's supposed to be near-perfect for high combo scores, and is relatively accurate for most scores so you should be able to ignore it when estimating what a play imported from stable would give, what matters the most is the acc change as it will probably partially compensate the absence of slider-acc.

We need data for stable vs lazer classic vs lazer to see:
- What woulf be good multiplier for stable scores vs lazer (both top 10 and random scores, on different star ratings)
- Can classic plays made on lazer use the same multiplier or should it be lower (probably should be lower, according to what we said above)

**Reply by Zyfarok** [reply]
Regarding the combo portion code: If I were you, I would just obtain the `ComboScore` by taking `TotalScore - AccScore` and then a change of acc is simply `ComboScore = ComboScore * newAcc / oldAcc`. This would have been very fine for those kind of estimations and does not require any replay.

**Reply by sz-da** [reply]
> Regarding the combo portion code: If I were you, I would just obtain the `ComboScore` by taking `TotalScore - AccScore` and then a change of acc is simply `ComboScore = ComboScore * newAcc / oldAcc`. This would have been very fine for those kind of estimations and does not require any replay.

That works too, but I need the replay to calculate the 100s from slider acc anyways, so it wouldn't have made a big difference.

I've updated the code to also include comparison with stable scores, and currently rerunning through all the replays (luckily I cached all the replays so this will take much less time).

**Reply by sz-da** [reply]
> * What would be good multiplier for stable scores vs lazer (both top 10 and random scores, on different star ratings)

For the top 10, the average barely changes ([sheet](https://docs.google.com/spreadsheets/d/15yrzNBWX4ubDYLHxfvSkqwrEWJARbjLPqMSKIMYhgp4/edit?usp=sharing))
classic avg: `0.971`
stable avg: `0.972`

For qualified maps, the average does increase a bit when using stable scores ([sheet](https://docs.google.com/spreadsheets/d/1j1TnLYKnMhOvl2glA_JqMPUyuZLZ85ASxktuc9RHZdA/edit?usp=sharing))
classic avg: `0.904`
stable avg: `0.921`

> * Can classic plays made on lazer use the same multiplier or should it be lower (probably should be lower, according to what we said above)

The multiplier does differ a bit for low acc scores, but imo it's not enough to really justify using a different multiplier. Especially when the multiplier is already pretty off for low acc scores.

**Comment by Newbie1st** [comment]
I would like to suggest a dynamic multiplier based on the slider percentage, ranging from 1x for maps with barely any sliders to something like 0.9x for maps which contain almost exclusively sliders. In my opinion, it's incorrect to punish scores where slider-acc doesn't play a role whatsoever, while the current 0.96x multiplier might not be enough to represent the difference in difficulty on slider-heavy maps. Can you please check the stats based on the slider percentage?

**Reply by Newbie1st** [reply]
Also a little bit off-topic, but from a mania-player standpoint, this approach would make even more sense, due to the 1m score hardcap and LNs being even more challenging to acc, so there is also a uniformity between gamemodes argument.

**Reply by ziv-vy** [reply]
This was rejected for being too complicated.
